asmcomp/cmmgen.ml

```diff
@@ -1307,21 +1350,44 @@ let rec is_unboxed_number e =
       join (is_unboxed_number e1) e2
   | _ -> No_unboxing
 
-let subst_boxed_number box_fn unbox_fn boxed_id unboxed_id box_chunk box_offset exp =
+let box_chunk = function
+  | Boxed_float -> Double_u
+  | Boxed_integer Pint32 -> Thirtytwo_unsigned
+  | Boxed_integer _ -> Word_int
+
+let box_offset = function
+  | Boxed_float -> 0
+  | Boxed_integer _ -> size_addr
+
+let subst_boxed_number boxed_number boxed_id unboxed_id exp =
   let rec subst = function
-    | Cvar id when Ident.same id boxed_id -> box_fn (Cvar unboxed_id)
+    | Cvar id when Ident.same id boxed_id ->
+        box_number boxed_number (Cvar unboxed_id)
     | Clet(id, arg, body) -> Clet(id, subst arg, subst body)
     | Cassign(id, arg) when Ident.same id boxed_id ->
-        Cassign(unboxed_id, subst(unbox_fn arg))
+        Cassign(unboxed_id, subst(unbox_number boxed_number arg))
     | Cassign(id, arg) -> Cassign(id, subst arg)
-    | Ctuple argv -> Ctuple(List.map subst argv)
     | Cop(Cload chunk, [Cvar id])
-      when Ident.same id boxed_id && chunk = box_chunk && 0 = box_offset ->
+      when Ident.same id boxed_id &&
+           chunk = box_chunk boxed_number &&
+           0 = box_offset boxed_number ->
         Cvar unboxed_id
     | Cop(Cload chunk, [Cop(Cadda, [Cvar id; Cconst_int ofs])])
-      when Ident.same id boxed_id && chunk = box_chunk && ofs = box_offset ->
+      when Ident.same id boxed_id
+           && chunk = box_chunk boxed_number &&
+           ofs = box_offset boxed_number ->
         Cvar unboxed_id
     | Cop(op, argv) -> Cop(op, List.map subst argv)
+    (* 64 bits integer on 32 bit target *)
```

Comment:
If we compare the output of `-dcmm` we'll need one reference file for 32 bits and one for 64 bits. It might to be a bit annoying to update these files when the cmm changes.

I think it's enough to write a small programs that fires this optimization and check that nothing was allocated.